### PR TITLE
Rework grouping of import files

### PR DESCRIPTION
**Pull request by czbarth**
Renamed the possible actions and the method that builds the import group. 
The import group is now provided by the auto importer and contains the primary file already. With that it's more unlikely that the `IImportFormat` creates an unexpected import group.
Added an exception that is thrown when reading data from a `IFileSource `failed. That exception is handled by the AutoImporter as an error.

### CODE REVIEW

File: src/PiWeb.Import.Sdk/ImportData/AdditionalFileSourceDataItem.cs [hunk 1]
```diff
@@ -26,7 +31,7 @@ public sealed class AdditionalFileSourceDataItem : AdditionalDataItem
 {
     #region members
 
-    private readonly IFileSource _FileSource;
+    private readonly IFileSource _fileSource;
```
Comment (czkraft) [inline src/PiWeb.Import.Sdk/ImportData/AdditionalFileSourceDataItem.cs:34]: Why suddenly lowercase?

File: src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportGroupExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+/// Provides methods to work with <see cref="IImportGroup"/> more easily.
+/// </summary>
+public static class ImportGroupExtensions
+{
+    #region methods
+
+    /// <summary>
+    /// Checks whether <paramref name="importGroup"/>'s <see cref="IImportGroup.PrimaryFile"/> has the requested <paramref name="extension"/>.
+    /// </summary>
+    /// <param name="importGroup">The <see cref="IImportGroup.PrimaryFile"/> to check.</param>
```
Comment (czkraft) [inline src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportGroupExtensions.cs:31]: `The group containing the primary file to check.`

File: src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportGroupExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    /// <param name="extension">The file extension (including the period ".").</param>
+    /// <returns><c>true</c> if the <see cref="IImportGroup.PrimaryFile"/> has the extension; otherwise <c>false</c>.</returns>
+    public static bool HasExtension(this IImportGroup importGroup, string extension)
+    {
+        if (!extension.StartsWith('.'))
+            extension = '.' + extension;
+
+        return importGroup.PrimaryFile.Name.EndsWith(extension, StringComparison.OrdinalIgnoreCase);
```
Comment (czkraft) [inline src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportGroupExtensions.cs:39]: I'm not sure whether we need those extension methods...
But we should at least use the FileSourceExtension here instead of implementing the comparison again:
`return importGroup.PrimaryFile.HasExtension(extension);`

File: src/PiWeb.Import.Sdk/Modules/ImportFormat/IImportGroup.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    /// The list of files recognized by the import format including <see cref="PrimaryFile"/>.
+    /// These files will not be offered to other import formats.
+    /// </summary>
+    ICollection<IFileSource> AllFiles { get; }
+
+    /// <summary>
+    /// Adds the <paramref name="fileSource"/> to this <see cref="IImportGroup"/>. 
+    /// These files will not be offered to other import formats. 
+    /// </summary>
+    void AddFile(IFileSource fileSource);
```
Comment (czkraft) [inline src/PiWeb.Import.Sdk/Modules/ImportFormat/IImportGroup.cs:47]: Don't we want to allow deletion of file sources (aside the primary file)?

File: src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportAction.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    Discard,
+
+    /// <summary>
+    /// <see cref="IImportGroup.PrimaryFile"/> was recognized and the <see cref="IImportGroup"/> is ready to be imported but there is
+    /// information missing. This information might be available later.
+    /// The files in <see cref="IImportGroup.AllFiles"/> will not be offered to another import format in the ongoing import run, but they
+    /// will be available in the next import run again. 
+    /// If the retention period of the files in <see cref="IImportGroup.AllFiles"/> is expired the <see cref="IImportGroup"/> is imported
+    /// with the available information.
+    /// </summary>
```
Comment (czkraft) [inline src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportAction.cs:44]: I would rephrase it a little:
```
   /// <summary>
    /// <see cref="IImportGroup.PrimaryFile"/> was recognized and the <see cref="IImportGroup"/> contains enough information to be imported at once,
    /// but can also wait for additional information that might appear later.
    /// So the <see cref="IImportGroup"/> should only be imported if the retention period of the files in <see cref="IImportGroup.AllFiles"/> is expired.
    /// If the retention period is not expired, the files in <see cref="IImportGroup.AllFiles"/> will not be imported in the ongoing import run
    /// but will be available again in the next import run. 
    /// </summary>
```

File: src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportAction.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    RetryOrImport,
+
+    /// <summary>
+    /// <see cref="IImportGroup.PrimaryFile"/> was recognized and the <see cref="IImportGroup"/> is ready to be imported but there is
+    /// information missing. This information might be available later.
+    /// The files in <see cref="IImportGroup.AllFiles"/> will not be offered to another import format in the ongoing import run, but they
+    /// will be available in the next import run again.
+    /// If the retention period of the files in <see cref="IImportGroup.AllFiles"/> is expired the <see cref="IImportGroup"/> is discarded
+    /// and the files are removed from the import folder.
+    /// </summary>
```
Comment (czkraft) [inline src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportAction.cs:54]:
```
   /// <summary>
    /// <see cref="IImportGroup.PrimaryFile"/> was recognized and the <see cref="IImportGroup"/> contains not enough information to be imported at once,
    /// but can wait for additional information that might appear later.
    /// So the <see cref="IImportGroup"/> should only be discarded if the retention period of the files in <see cref="IImportGroup.AllFiles"/> is expired.
    /// If the retention period is not expired, the files in <see cref="IImportGroup.AllFiles"/> will not be discarded in the ongoing import run
    /// but will be available again in the next import run. 
    /// </summary>
```

Comment (czbarth) [reply src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportAction.cs]: I changed the first part of the text al little bit and used your suggestion for the second part.

File: src/PiWeb.Import.Sdk/Modules/ImportFormat/IImportGroup.cs (hunk 4 above)
Comment (czbarth) [reply src/PiWeb.Import.Sdk/Modules/ImportFormat/IImportGroup.cs:47]: I skipped it, because it was not needed yet. But as it is not much work I added it now nevertheless.

File: src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportGroupExtensions.cs (hunk 3 above)
Comment (czbarth) [reply src/PiWeb.Import.Sdk/Modules/ImportFormat/ImportGroupExtensions.cs]: You are right

**Review by czkraft — APPROVED** [review]
